text/0000-pub-sub.md

```diff
@@ -0,0 +1,171 @@
+
+- Start Date: 2019-06-20
+- RFC PR: 
+- FOLIO Issue: 
+
+# PubSub support in FOLIO (module implementation)
+
+## Summary
+This RFC proposes to add event-driven architecture capabilities to the FOLIO platform. Specifically, it is a general purpose publisher-subscriber (`pubsub`) implementation.
+
+## Motivation
+
+The Folio Platform consists of a set of microservices which are organized around a central proxying API Gateway. Folio modules implement these microservices through RESTful API interfaces. In order to implement complex library workflows it is necessary to distribute the work to individual modules to perform the specialized tasks.
+
+The approach described here is to provide a messaging system whereby modules publish relevant data when ready, for other modules to consume and do their part. Modules which participate in a particular workflow can subscribe to the relevant messages. The advantage of this particular solution is that it fits well with the microservices architecture and preserves the independence of the modules. Each module plays its part but remains unaware of any role played by other modules. It does not require a centralized workflow controller.
+
+## Detailed Explanation/Design
+
+### Requirements
+#### Functional Requirements
+- A business module (back-end) can define a set of events it can send (publish). It can also declare and receive (subscribe to) event of different types.
+
+- An “Event Descriptor” must be provided to define new event types.
+  1. The event definition produces a unique “Event Type”
+  2. The “Event Descriptor” must have a detailed description for the event type and payloads which can be received within events of this type.
+  3. “Event Descriptor” does not support versioning. To update an existing event declaration, a new "Event Type" must be defined with a new "Event Descriptor". However, it is recommended that the updated "Event Type" maintain the same name followed by an incremented version number suffix. The description field of the new event type should be used to explain the difference between the two related “Event Descriptors”. 
+  4. The “Event Descriptor” must define a default time-to-live (TTL) for events of this type.
```

Comment:
What should happen when a message exceeds the defined time to live? Is it deleted? Does delivery to subscribers cease?

Reply:
In this proposal a message exceeding TTL is dead and will not be delivered to any subscribers. Whether and how it is deleted is an implementation detail.

Reply:
Sure, agree that deletion is an implementation detail (I like to ask questions multiple ways in case on wasn’t very clear). 

Thank you for clarifying the expectation that no further deliveries will occur when the TTL has been exceeded.